pyproject_metadata/__init__.py

```diff
@@ -132,10 +132,11 @@ class _SmartMessageSetter:
     def __setitem__(self, name: str, value: str | None) -> None:
         if not value:
             return
-        if '\n' in value:
-            msg = f'"{name}" should not be multiline; indenting to avoid breakage'
+        lines = value.splitlines()
+        if len(lines) > 1:
+            msg = f'"{name}" should not be multiline; joining to avoid breakage'
             warnings.warn(msg, ConfigurationWarning, stacklevel=2)
-            value = value.replace('\n', '\n        ')
+            value = ' '.join(lines)
```

Comment:
I haven't checked what `compat32` does, but historically there is a limit on the length of header lines. And the `License` field can contain the whole text of the license (that is what the `project.license.file` field is for, after all) which may be very long. I don't know if that is enough to bump into the length limit.

I'm very surprised that there isn't a way to have utf8 and multi-line fields simultaneously. I don't see why it should not be possible.

Reply:
This also brings up the issue that the warning added for multi-line fields is emitted for the `License` field. Which, as I wrote above, often expected to contain multi line text. I am no so sure that the warning is a good idea after all. What is the reason for the warning? Evidently I was very wrong when I suggested that we may not need support for multi-line fields

Reply:
> I haven't checked what `compat32` does, but historically there is a limit on the length of header lines

To clarify, this is not about the serialization on our side (that we can tweak at will) bur rather about the deserialiazation. It would be bad if the fields would be truncated when they are read.

Reply:
> I'm very surprised that there isn't a way to have utf8 and multi-line fields simultaneously.

Note to myself: the `email.policy.Policy` `fold` and `refold_source` seem to be related to this.

Reply:
~~Adding unicode handling to fold and fold_binary seems to have done it.~~ Ah, wait a minute, https://inspector.pypi.io/project/fabio/2024.9.0/packages/92/6d/4c17e4d694d97168bb7d780bf10fb45c39f2dcef4aa5790f1b794ef9e09e/fabio-2024.9.0-cp39-cp39-win_amd64.whl/fabio-2024.9.0.dist-info/METADATA does have the escape sequences. Maybe the simplest case of lying about utf8 would work.

Reply:
I think the escape sequences work more by accident than by design. The core metadata design says that `METADATA` should be utf8 encoded. Thus the easiest is to use utf8 encoding and not the escape sequences.

We could check what the venerable setuptools does, but I don't know how relevant that is.

Reply:
It also says you should be able to read it with `email` using `compat32`. 🙄

Reply:
Though `message_from_string` is fine, it's just `message_from_bytes` that fails.

Reply:
What's the point in specifying it should be utf8 if then any non-ascii character results in an error? The specification seems contradictory.

Reply:
I've the impression that `email.message.Message` and `email.message.EmailMessage` do things a bit differently in this regard. Which one is to be used?

Reply:
The problem is `message_from_bytes` use ascii + escapes. But `message_from_string` will work with unicode strings, so maybe the spec expects everyone reading it to use that.

The thing that confuses me is why the previous METADATA files had escape sequences in them rather than unicode.

Reply:
I think this restores the old behavior, and uses unicode. We could try it as `0.9.0b2` and see.

Reply:
Moved to Message. I thought EmailMessage was a replacement, but it seems like it's not as compatible with Compat32 as Message is, and Message is not deprecated as far as I can tell.

So when is that JSON metadata coming? :D